src/gather_classes.rs

```diff
@@ -257,80 +255,131 @@ fn update_report(
             debug!("skipping transaction: `{transaction}`");
             continue;
         }
-        let calls = get_calls(root_call).context("Failed to get calls from root_call")?;
-        for call in calls {
-            let maybe_kvp = get_tuple_from_call(call);
-            match maybe_kvp {
-                Ok(kvp) => {
-                    report.add_fields(kvp.1, kvp.0);
-                    updates += 1;
-                }
-                Err(err) => {
-                    debug!("Failed to extract tuple with error: {err}");
-                }
-            }
+        let calls =
+            get_calls_with_count(root_call).context("Failed to get calls from root_call")?;
+        for ((entry_point, class_hash), count) in calls {
+            report.update_count(class_hash, entry_point, count);
+            updates += count;
         }
     }
     assert!(report.blocks.insert(block_num));
     Ok(updates)
 }
 
-// Converts recursive calls to value.get("calls") to a boxed iterator.
-fn get_calls<'a>(obj: &'a Value) -> Result<Box<dyn Iterator<Item = &Value> + 'a>, anyhow::Error> {
+/// Merges two lists of calls into a single list of calls. We only consider a [CallWithCount] if
+/// it exists in both lists.
+///
+/// This means that if native is empty, then no calls will be added to the merged list.
+fn merge_calls_with_count(
+    base_calls: Vec<CallWithCount>,
+    native_calls: Vec<CallWithCount>,
+) -> Vec<CallWithCount> {
+    let mut merged_map: HashMap<CallKey, usize> = HashMap::new();
+
+    let native_keys: HashSet<CallKey> = native_calls
+        .iter()
+        .map(|((ep, ch), _)| (*ep, *ch))
+        .collect();
+    for (call_key, base_count) in base_calls {
+        if native_keys.contains(&call_key) {
+            merged_map
+                .entry(call_key)
+                .and_modify(|count| *count += base_count)
+                .or_insert(base_count);
+        }
+    }
+
+    merged_map.into_iter().collect()
```

Comment:
Why are you using a vector of tuples, instead of a HashMap?

Signaling it in this line, but it applies to the whole solution. Why don't just deal with a "Map" directly?

Reply:
Initially, I kept it as a vector because it was easier to combine instead of creating multiple hashmaps. But now that I'm not chaining vecs, I think it's better to use HashMaps